### DISCUSSION TITLE
Writing unit tests for custom generators and types

### DISCUSSION THREAD

**Discussion by LeoAdamek** [doctrine/doctrine2#8647] [Support Questions] [answered]
My application makes use of a custom id type which maps to a time-based identity, it also makes use of a custom generator for these values. What's the best way to write tests for this? e.g. Initializing the EntityManager etc (within Symfony Flex.)

~~~~php
namespace App\Tests\Support\Database\Identity;


use App\Support\Database\Identity\Type;
use Doctrine\DBAL\Platforms\PostgreSQL100Platform;
use PHPUnit\Framework\TestCase;

final class TypeTests extends TestCase
{

    public function testGetSqlDeclaration(): void
    {
        $type = new Type();
        $this->assertEquals('ksuid', $type->getSQLDeclaration([], new PostgreSQL100Platform()));
    }


}
~~~~

This works well enough to handle the SQL declaration, but to test serialization and deserialization of DB values, or the ID generator which needs an active EntityManager, how would I approach this?

**Comment by beberlei** [comment, marked as answer]
For an integration test, Use the KernelTestCase from Symfony and fetch the entity manager service from the container.

**Reply by LeoAdamek** [reply]
Thanks, using `KernelTestCase` to have a container to get the service from was exactly what I needed.
